internal/testing/e2e/backfill_test.go

```diff
@@ -318,13 +303,69 @@ func TestBackfillGenerationMismatch(t *testing.T) {
 		Config:  om.MMFConfigGRPC(),
 		Profile: &pb.MatchProfile{},
 	})
-	require.Nil(t, err)
+	require.NoError(t, err)
 
 	resp, err := stream.Recv()
 	require.Nil(t, resp)
 	require.Equal(t, io.EOF, err)
 }
 
+func TestCleanUpBackfills(t *testing.T) {
+	ctx := context.Background()
+	om := newOM(t)
+
+	t1, err := om.Frontend().CreateTicket(ctx, &pb.CreateTicketRequest{Ticket: &pb.Ticket{}})
+	require.NoError(t, err)
+	require.NotNil(t, t1)
+
+	b1, err := om.Frontend().CreateBackfill(ctx, &pb.CreateBackfillRequest{Backfill: &pb.Backfill{
+		SearchFields: &pb.SearchFields{
+			StringArgs: map[string]string{
+				"search": "me",
+			},
+		}}})
+	require.NoError(t, err)
+	require.NotNil(t, b1)
+
+	m := &pb.Match{
+		MatchId:  "1",
+		Tickets:  []*pb.Ticket{t1},
+		Backfill: b1,
+	}
+
+	om.SetMMF(func(ctx context.Context, profile *pb.MatchProfile, out chan<- *pb.Match) error {
+		out <- m
+		return nil
+	})
+
+	om.SetEvaluator(func(ctx context.Context, in <-chan *pb.Match, out chan<- string) error {
+		p, ok := <-in
+		require.True(t, ok)
+		require.True(t, proto.Equal(p, m))
+		_, ok = <-in
+		require.False(t, ok)
+
+		out <- m.MatchId
+		return nil
+	})
+
+	// wait until backfill is expired, then try to get it
+	time.Sleep(pendingReleaseTimeout * 3)
+
+	// statestore.CleanupBackfills is called at the end of each syncronizer cycle after fetch matches call, so expired backfill will be removed
+	stream, err := om.Backend().FetchMatches(ctx, &pb.FetchMatchesRequest{
+		Config:  om.MMFConfigGRPC(),
+		Profile: &pb.MatchProfile{},
+	})
+	require.NoError(t, err)
+
+	_, err = stream.Recv()
+	require.Error(t, err)
+	e, ok := status.FromError(err)
+	require.True(t, ok)
+	require.Contains(t, e.Message(), "error(s) in FetchMatches call. syncErr=[failed to handle match backfill: 1: rpc error: code = NotFound desc = Backfill id:")
```

Comment:
Curious about the test logic here. Given your sleep is even before the FetchMatches call - how is that sleep causing backfill expiration? I would expect FetchMatches call to trigger MMF and have a backfill generated and so shouldnt the sleep be after the call?

Reply:
I'm pretty sure the wait prior to FetchMatches is valid. The Backfill is created right after the call to CreateTicket at the start of the test. The Backfill is acknowledged right when it is created, and that sets its expiration timestamp, so waiting prior to calling FetchMatches can cause the Backfill to age.
I think this test mirrors the case where a game server has a player leave a game, and it creates the Backfill on its own. Maybe we need another test where the MMF is not able to completely fill a Match, and requests a Backfill to be created?

Reply:
Thanks. That case makes sense - yes I was confusing it with the MMF created backfill case - and +1 to having a test case to cover that.